### ISSUE TITLE
Nix overlays do not work as documented in nixpkgs manual

### ISSUE THREAD

**Issue by seppeljordan** [NixOS/nixpkgs#23016]
## Issue description

nixpkgs overlays don't work as described in the manual.  It says:

> For the second and third options, the directory should contain Nix expressions defining the overlays. Each overlay can be a file, a directory containing a default.nix, or a symlink to one of those. The expressions should follow the syntax described in Section 11.2, “Overlays Layout”.

In the PR for nixpkgs overlays it says:

> I've changed $NIXPKGS_OVERLAYS to <nixpkgs-overlays>, and ~/.nixpkgs/overlays to ~/.config/nixpkgs/overlays in order to not add more non-XDG-compliant paths.

Now if I create a file ~/.config/nixpkgs/overlays/example.nix with modifications to nixpkgs in it, these modifications get applied.  When I create a file ~/.config/nixpkgs/overlays/10-example/default.nix with the same contents, it does not get applied.

### Steps to reproduce

The following works

    $ mkdir -p ~/.config/nixpkgs/overlays/example/
    $ echo 'self: super: { helloo = self.hello; }' > ~/.config/nixpkgs/overlays/example.nix
    $ nix-env -f '<nixpkgs>' -qaA 'helloo'

The following does not work

    $ mkdir -p ~/.config/nixpkgs/overlays/example/
    $ echo 'self: super: { helloo = self.hello; }' > ~/.config/nixpkgs/overlays/example/default.nix
    $ nix-env -f '<nixpkgs>' -qaA 'helloo'

## Technical details

* System: nixos
* Nix version: nix-env (Nix) 1.11.6
* Nixpkgs version: "17.03pre101267.a9584c9"

**Comment by teh** [comment]
@seppeljordan - the config has changed:
https://github.com/NixOS/nixpkgs/commit/7dacca324d9e86fe7f7fad450ec833c746f21415

cc @nbp - I can reproduce even without going through a file though:

```
NIX_PATH=nixpkgs-base=/home/tom/.nix-defexpr/channels/nixpkgs:nixpkgs=$PWD/x.nix nix-env -iA nixpkgs.hello2
```

fails. With `x.nix`: 
```
let ov = self: super: {
    hello2 = self.hello;
};
in (import <nixpkgs-base> { overlays = [ ov ]; })
```

**Comment by teh** [comment]
Even more weird, it works sometimes (see comments):

```
let
  ov = self: super: {
    hello2 = super.hello; # does not work
    hello3 = super.python; # works?
  };
  pkgs = import <nixpkgs> {};
in import pkgs.path { overlays = [ov]; }
```

**Comment by dtzWill** [comment]
This seems related: https://github.com/NixOS/nixpkgs/commit/86fe7a40acc0e68e60905798b2d692c2e9a6c33e#commitcomment-20704454

**Comment by nbp** [comment]
cc @edolstra , what about never asking for review and allow everybody to push to the repository, even for small commits?

**Comment by nbp** [comment]
This should be fixed by #23102
